processor/normalizesumsprocessor/processor.go

```diff
@@ -0,0 +1,369 @@
+	var errors []error
+
+	ilms := rms.InstrumentationLibraryMetrics()
+	for j := 0; j < ilms.Len(); j++ {
+		ilm := ilms.At(j).Metrics()
+		for k := 0; k < ilm.Len(); k++ {
+			metric := ilm.At(k)
+			if metric.DataType() == pdata.MetricDataTypeIntSum || metric.DataType() == pdata.MetricDataTypeDoubleSum {
+				err := nsp.transformMetric(SumMetrics{MetricName: metric.Name()}, rms.Resource(), &metric, &ilm)
+				if err != nil {
+					errors = append(errors, err)
+				}
+			}
+		}
+	}
+
+	return errors
+}
+
+//func findMetric(name string, rms pdata.ResourceMetrics) (*pdata.Metric, *pdata.MetricSlice) {
+func (nsp *NormalizeSumsProcessor) transformMetric(transform SumMetrics, resource pdata.Resource, metric *pdata.Metric, slice *pdata.MetricSlice) error {
+	switch t := metric.DataType(); t {
+	case pdata.MetricDataTypeIntSum:
+		dataPointsRemaining, processingErr := nsp.processIntSumMetric(slice, transform, resource, metric)
+		if processingErr != nil {
+			return processingErr
+		} else {
+			cleanUpMetricSlice(transform, metric, slice, dataPointsRemaining)
+		}
+	case pdata.MetricDataTypeDoubleSum:
+		dataPointsRemaining, processingErr := nsp.processDoubleSumMetric(slice, transform, resource, metric)
+		if processingErr != nil {
+			return processingErr
+		} else {
+			cleanUpMetricSlice(transform, metric, slice, dataPointsRemaining)
+		}
+	default:
+		return fmt.Errorf("Data Type not supported %s", metric.DataType())
+	}
+
+	return nil
+}
+
+func cleanUpMetricSlice(transform SumMetrics, metric *pdata.Metric, slice *pdata.MetricSlice, dataPointsRemaining int) {
+	// If there is meaningful data to send and we are renaming the metric,
+	// add the already renamed metric to the slice
+	if dataPointsRemaining > 0 && transform.NewName != "" {
+		newMetric := pdata.NewMetric()
+		metric.CopyTo(newMetric)
+		newMetric.SetName(transform.NewName)
+
+		slice.Append(*&newMetric)
+	}
+
+	// If there are no remaining data points after removing restart/start
+	// points, or this metric was renamed, remove this metric from the slice
+	if dataPointsRemaining == 0 || transform.NewName != "" {
+		metricSliceRemoveElement(slice, transform.MetricName)
+	}
+}
+
+func (nsp *NormalizeSumsProcessor) processIntSumMetric(slice *pdata.MetricSlice, transform SumMetrics, resource pdata.Resource, metric *pdata.Metric) (int, error) {
+	dps := metric.IntSum().DataPoints()
+	for i := 0; i < dps.Len(); {
+		reportData := nsp.processIntSumDataPoint(dps.At(i), resource, metric)
+
+		if reportData == false {
+			intRemoveAt(&dps, i)
+			continue
+		}
+		i++
+	}
+
+	return dps.Len(), nil
+}
+
+func (nsp *NormalizeSumsProcessor) processIntSumDataPoint(dp pdata.IntDataPoint, resource pdata.Resource, metric *pdata.Metric) bool {
+	metricIdentifier := dataPointIdentifier(resource, metric, dp.LabelsMap())
+
+	start := nsp.history[metricIdentifier]
+	// If this is the first time we've observed this unique metric,
```

Comment:
I believe this logic already exists in OT in the prometheus receiver; we should try to avoid duplicating that code here. (But also, it would be good to have some understanding of why nginxreceiver needs this in the first place - it sounds like the metrics it's emitting are already broken.)

Reply:
The metrics being emitted from the nginxreceiver have no start time reference due to the way metrics are exposed by nginx. In order to present meaningful data, we treat the first point as the start reference. You lose the information of the full counter value, but retain the information of how the counter value is growing over time.

Reply:
Yeah, I'm just trying to argue against duplicated code - it seems like this is something that should (eventually) be fixed in the nginxreceiver directly, using the same logic and code as the prometheus receiver already does. There's lots of subtle edge cases to handle, as I'm sure you've discovered, so it's better to fix it once right and carefully.

(or is this already a fork of the code in the prometheus receiver?)

Reply:
Let me be clearer on this:

I know from personal experience that this is very subtle logic that is hard to get right in all the edge cases.

AFAIK this logic already exists in the OT collector repo, as the type [`MetricsAdjuster`](https://github.com/open-telemetry/opentelemetry-collector/blob/6e5beaf43b325e63ec6f1e864d9746a0d051cc35/receiver/prometheusreceiver/internal/metrics_adjuster.go#L187). Long-term I'd like to see this logic exist in exactly one place in OT, with a robust set of tests, called by all the different receivers/processors/exporters that need it. I think it's more likely that the logic will be correct if we do that.

I'm okay with making cleaning this up a TODO for now, since I don't want to block on this, but I don't want to end up in this state long-term. Can you add a TODO comment to merge this with `MetricsAdjuster` in a single common location? Then I'll do my best to review the current logic for any subtle bugs, but as I said, I know how hard it is to get right.

Reply:
That's reasonable. In a quick look, it appears that MetricsAdjuster in its current state does not attempt to use the initial collection as a zero value for a sum metric, but otherwise it's definitely a lot of duplication and shared concepts. Happy to add the TODO .